src/main/java/net/fabricmc/loom/util/fmj/FabricModJsonHelpers.java

```
@@ -24,39 +24,54 @@
 
 package net.fabricmc.loom.util.fmj;
 
+import java.io.File;
 import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
+import net.fabricmc.loom.api.LoomGradleExtensionAPI;
 import org.gradle.api.Project;
 import org.gradle.api.tasks.SourceSet;
 
 import net.fabricmc.loom.LoomGradleExtension;
 import net.fabricmc.loom.util.gradle.SourceSetHelper;
+import org.jetbrains.annotations.Nullable;
 
 public class FabricModJsonHelpers {
-	// Returns a list of Mods found in the provided project's main or client sourcesets
+	/**
+	 * Returns the list of mods provided by either {@link LoomGradleExtensionAPI#getFabricModJsonPath()}
+	 * or {@code fabric.mod.json} in main or client resources.
+	 */
 	public static List<FabricModJson> getModsInProject(Project project) {
-		final LoomGradleExtension extension = LoomGradleExtension.get(project);
+		LoomGradleExtension extension = LoomGradleExtension.get(project);
+		var overrideFile = extension.getFabricModJsonPath().getAsFile();
```

I think this can just return early if the override file is present. If it is it should try to read the FMJ and throw if it fails to read the FMJ. We shouldnt sliently hide errors from the user like this.